src/tree/treemath.rs

```diff
@@ -69,40 +72,73 @@ pub(crate) fn right(index: NodeIndex, size: LeafIndex) -> Result<NodeIndex, Tree
     Ok(NodeIndex::from(r))
 }
 
+// The parent here might be beyond the right edge of the tree.
 pub(crate) fn parent_step(x: usize) -> usize {
     let k = level(NodeIndex::from(x));
     let b = (x >> (k + 1)) & 0x01;
     (x | (1 << k)) ^ (b << (k + 1))
 }
 
 pub(crate) fn parent(index: NodeIndex, size: LeafIndex) -> Result<NodeIndex, TreeMathError> {
+    node_in_tree(index, size)?;
+    unsafe_parent(index, size)
+}
+
+// This function is only safe to use if index <= size.
+// If this is not checked before calling the function, `parent` should be used.
+fn unsafe_parent(index: NodeIndex, size: LeafIndex) -> Result<NodeIndex, TreeMathError> {
```

Comment:
If this function returns `Result` anyway, why not just have it return an error if the input index is outside of the tree?

Reply:
This is a pure "performance improvement". It's avoiding the check when we know that it was done before. (I don't think we'd notice the performance difference...)